src/converters.js

```diff
@@ -814,6 +799,66 @@ function generateLiInUl( modelItem, conversionApi ) {
 	return viewItem;
 }
 
+// Helper function that converts children of a given `<li>` view element into corresponding model elements.
+// The function maintains proper order of elements if model `listItem` is split during the conversion
+// due to block children conversion.
+//
+// @param {module:engine/model/element~Element} listItemModel List item model element to which converted children will be inserted.
+// @param {Iterable.<module:engine/view/node~Node>} viewChildren View elements which will be converted.
+// @param {Object} conversionApi Conversion interface to be used by the callback.
+// @returns {module:engine/model/position~Position} Position on which next elements should be inserted after children conversion.
+function viewToModelListItemChildrenConverter( listItemModel, viewChildren, conversionApi ) {
+	let lastListItem = listItemModel;
+	let nextPosition = ModelPosition.createAfter( lastListItem );
+
+	// Check all children of the converted `<li>`. At this point we assume there are no "whitespace" view text nodes
+	// in view list, between view list items. This should be handled by `<ul>` and `<ol>` converters.
+	for ( const child of viewChildren ) {
+		// If this is a view list element, we will convert it after last `listItem` model element.
+		if ( child.name == 'ul' || child.name == 'ol' ) {
+			nextPosition = conversionApi.convertItem( child, nextPosition ).modelCursor;
+		}
+		// If it was not a list view element it was a "regular" list item content. Convert it to a `listItem`.
+		else {
+			const result = conversionApi.convertItem( child, ModelPosition.createAt( lastListItem, 'end' ) );
+			const convertedChild = Array.from( result.modelRange )[ 0 ];
+
+			nextPosition = ModelPosition.createAt( lastListItem, 'end' );
```

Comment:
Isn't it the same as `result.modelCursor`? I honestly don't know, I think I haven't yet written a view-to-model converter after the refactor :P.